sdk/resourcemanager/Azure.ResourceManager.Core/src/ApiVersions.cs

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using Azure.ResourceManager.Resources;
+using System.Reflection;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+
+namespace Azure.ResourceManager.Core
+{
+    /// <summary>
+    /// A class representing Azure resource manager client options.
+    /// </summary>
+    public class ApiVersions
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ApiVersions"/> class.
+        /// </summary>
+        public ApiVersions()
+        {
+            BuildApiTable();
+        }
+
+        private Dictionary<string, string> _resourceToApiVersions = new Dictionary<string, string>();
+
+        private void BuildApiTable()
+        {
+            var methods = GetExtensionMethods();
+            foreach (var method in methods)
+            {
+                if (method.Name.EndsWith("RestApiVersions"))
+                {
+                    var apiObject = method.Invoke(null, new object[] { this });
+                    var properties = apiObject.GetType().GetProperties();
+                    foreach (var prop in properties)
+                    {
+                        if (typeof(ApiVersionsBase).IsAssignableFrom(prop.PropertyType))
+                        {
+                            var propVal = (ApiVersionsBase)prop.GetValue(apiObject);
+                            Console.WriteLine(prop.GetType());
+                            var key = propVal.ResourceType;
+                            _resourceToApiVersions.Add(key.ToString(), propVal);
```

Comment:
I am wondering if this is the right way to go about this.  As a customer I expect will do the following

var options = new ArmClientOptions();
options.ComputeApiVersions.VirtualMachineVersion = VirtualMachineVersion.2020_06_01;

how will your dictionary know that happened and stay up to date?

The SetApiVersion is needed for the case where I don't have Compute loaded, but when it is loaded I think the explicit method is more friendly?

Reply:
How do you image they add this field (ComputeApiVersions) to the ArmCleintOptions class? Via extensions inside compute? 

Can always just update the dictionary on the set if it is an extension.